django/contrib/gis/gdal/raster/band.py

```diff
@@ -67,3 +66,73 @@ def nodata_value(self):
         nodata_exists = c_int()
         value = capi.get_band_nodata_value(self.ptr, nodata_exists)
         return value if nodata_exists else None
+
+    @nodata_value.setter
+    def nodata_value(self, value):
+        """
+        Sets the nodata value for this band.
+        """
+        if not isinstance(value, (int, float)):
+            raise ValueError('Nodata value must be numeric.')
+        capi.set_band_nodata_value(self.ptr, value)
+
+    def datatype(self, as_string=False):
+        """
+        Returns the GDAL Pixel Datatype for this band.
+        """
+        dtype = capi.get_band_datatype(self.ptr)
+        if as_string:
+            dtype = GDAL_PIXEL_TYPES[dtype]
+        return dtype
+
+    def data(self, data=None, offset=None, size=None, as_buffer=False):
+        """
+        Reads or writes pixel values for this band. Blocks of data can
+        be accessed by specifying the with and height and the offset of
+        the desired block. The same specification can be used to update
+        parts of a raster by providing an array of values.
+
+        Allowed input data types are string, buffer, list, tuple and array.
+        """
+        if not offset:
+            offset = (0, 0)
+        if not size:
+            size = (self.width - offset[0], self.height - offset[1])
+
+        if any((x <= 0 for x in size)):
+            raise ValueError('Offset too big for this raster.')
+
+        if size[0] > self.width or size[1] > self.height:
+            raise ValueError('Size is larger than raster.')
+
+        # Create ctypes type array generator
+        ctypes_array = GDAL_TO_CTYPES[self.datatype()] * (size[0]*size[1])
+
+        if data is not None:
+            # Set write mode
+            access_flag = 1
+            # Instantiate ctypes array holding the input data
+            if isinstance(data, (str, buffer)):
+                data_array = ctypes_array.from_buffer_copy(data)
+            else:
+                data_array = ctypes_array(*data)
+        else:
+            # Set read mode
+            access_flag = 0
+            # Prepare empty ctypes array
+            data_array = ctypes_array()
+
+        # Access band
+        capi.band_io(self.ptr, access_flag, offset[0], offset[1],
+                     size[0], size[1], byref(data_array), size[0],
+                     size[1], self.datatype(), 0, 0)
+
+        # Return data as numpy array if possible, otherwise as list
+        if data is None:
+            if as_buffer:
+                return buffer(data_array)
+            elif numpy:
+                return numpy.frombuffer(
+                    data_array, dtype=numpy.dtype(data_array)).reshape(size)
+            else:
+                return list(data_array)
```

Comment:
I understand the concern, retrieving pixel values of a big raster fills up the memory quickly.

One simple option to aleviate this a bit (aside making two functions), is to warn about this in the documentation with a note/warning banner or so. The other option is to make the size argument a requirement, so that the user is very aware of the number of pixels he or she is retrieving.

I lean towards keeping just one function, but I am open to both solutions. A single function mirrors the underlying gdal function and splitting it into two would probably require making the dual one private and write two "wrappers" for get and set.